crates/ruff/src/rules/ruff/rules/mutable_defaults_in_dataclass_fields.rs

```diff
@@ -0,0 +1,207 @@
+/// ```
+#[violation]
+pub struct MutableDataclassDefault;
+
+impl Violation for MutableDataclassDefault {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Do not use mutable default values for dataclass attributes")
+    }
+}
+
+/// ## What it does
+/// Checks for function calls in dataclass defaults.
+///
+/// ## Why is it bad?
+/// Function calls are only performed once during definition time. The result
+/// gets reused in all created instances of the dataclass.
+///
+/// ## Examples:
+/// ```python
+/// from dataclasses import dataclass
+///
+/// def creating_list() -> list[]:
+///     return [1, 2, 3, 4]
+///
+/// @dataclass
+/// class A:
+///     mutable_default: list[int] = creating_list()
+///
+/// # also:
+///
+/// @dataclass
+/// class B:
+///     also_mutable_default_but_sneakier: A = A()
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from dataclasses import dataclass, field
+///
+/// def creating_list() -> list[]:
+///     return [1, 2, 3, 4]
+///
+/// @dataclass
+/// class A:
+///     mutable_default: list[int] = field(default_factory=creating_list)
+///
+/// @dataclass
+/// class B:
+///     also_mutable_default_but_sneakier: A = field(default_factory=A)
+/// ```
+///
+/// Alternatively, if you __want__ the shared behaviour, make it more obvious
+/// by assigning it to a module-level variable:
+/// ```python
+/// from dataclasses import dataclass
+///
+/// def creating_list() -> list[]:
+///     return [1, 2, 3, 4]
+///
+/// I_KNOW_THIS_IS_SHARED_STATE = creating_list()
+///
+/// @dataclass
+/// class A:
+///     mutable_default: list[int] = I_KNOW_THIS_IS_SHARED_STATE
+/// ```
+#[violation]
+pub struct FunctionCallInDataclassDefaultArgument {
+    pub name: Option<String>,
+}
+
+impl Violation for FunctionCallInDataclassDefaultArgument {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let FunctionCallInDataclassDefaultArgument { name } = self;
+        if let Some(name) = name {
+            format!("Do not perform function call `{name}` in dataclass defaults")
+        } else {
+            format!("Do not perform function call in dataclass defaults")
+        }
+    }
+}
+
+fn is_mutable_expr(expr: &Expr) -> bool {
+    matches!(
+        &expr.node,
+        ExprKind::List { .. }
+            | ExprKind::Dict { .. }
+            | ExprKind::Set { .. }
+            | ExprKind::ListComp { .. }
+            | ExprKind::DictComp { .. }
+            | ExprKind::SetComp { .. }
+    )
+}
+
+const ALLOWED_FUNCS: &[&[&str]] = &[&["dataclasses", "field"]];
+
+fn is_allowed_func(checker: &Checker, func: &Expr) -> bool {
+    checker
+        .ctx
+        .resolve_call_path(func)
+        .map_or(false, |call_path| {
+            ALLOWED_FUNCS
+                .iter()
+                .any(|target| call_path.as_slice() == *target)
+        })
+}
+
+pub fn function_call_in_dataclass_defaults(checker: &mut Checker, body: &[Stmt]) {
```

Comment:
Just for convenience, we'll often add a comment like:

```rust
/// RUF201
pub fn function_call_in_dataclass_defaults(checker: &mut Checker, body: &[Stmt]) {
  ...
}
```

Makes it easy to grep for the rule implementation given a code.

Reply:
done that :)